server/api.go

```diff
@@ -706,51 +714,57 @@ func (a *API) updateWhitelist(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "error reading whitelist", http.StatusBadRequest)
 		return
 	}
+	defer file.Close()
 
 	reader := csv.NewReader(file)
 	columns, err := reader.Read()
 	if err != nil || strings.ToLower(columns[0]) != "email" || len(columns) != 1 {
 		a.p.API.LogWarn("Error parsing whitelist csv header")
-		http.Error(w, "error parsing whitelist - bad csv header", http.StatusBadRequest)
+		http.Error(w, "error parsing whitelist - please check header and try again", http.StatusBadRequest)
 		return
 	}
 
-	records, err := reader.ReadAll()
-	if err != nil {
-		a.p.API.LogWarn("Error parsing whitelist csv data")
-		http.Error(w, "error parsing whitelist", http.StatusBadRequest)
-	}
-
-	var users []*model.User
-
-	for _, row := range records {
-		email := row[0]
-
-		user, err := a.p.API.GetUserByEmail(email)
-		if err != nil {
-			a.p.API.LogWarn("Error getting user")
-			http.Error(w, "error processing whitelist - could not find user with email: "+email, http.StatusInternalServerError)
-			return
+	var emails []string
+	var csvLineErrs []string
+	i := 1
+	for {
+		i++
+		row, readErr := reader.Read()
+		if readErr == io.EOF {
+			break
 		}
-
-		users = append(users, user)
+		if readErr != nil {
+			csvLineErrs = append(csvLineErrs, strconv.Itoa(i))
+		}
+		emails = append(emails, row[0])
 	}
 
-	if err := a.p.store.DeleteWhitelist(); err != nil {
-		a.p.API.LogWarn("Error deleting whitelist")
-		http.Error(w, "error processing whitelist", http.StatusInternalServerError)
+	if len(csvLineErrs) > MaxUpdateWhitelistCsvParseErr {
```

Comment:
I like the idea behind these changes, but I'd move some things between layers. As we're already iterating through the emails in the form of the CSV lines, I'd check them directly here through `a.p.API.GetUserByEmail(email)`, and only move forward to the store layer if all of them are found to make this operation atomic. If a read from the CSV fails, I would directly return an error as something is very wrong in that case. I'd remove the `MaxUpdateWhitelistCsvParseErr` threshold as well, as it should be an all of nothing, one failed email or CSV line read should cause an error for the whole import.